shared/contracts/BaseTemplate.sol

```diff
@@ -0,0 +1,191 @@
+pragma solidity 0.4.24;
+
+import "@aragon/apps-agent/contracts/Agent.sol";
+import "@aragon/apps-vault/contracts/Vault.sol";
+import "@aragon/apps-voting/contracts/Voting.sol";
+import "@aragon/apps-finance/contracts/Finance.sol";
+import "@aragon/apps-token-manager/contracts/TokenManager.sol";
+import "@aragon/apps-shared-minime/contracts/MiniMeToken.sol";
+
+import "@aragon/os/contracts/acl/ACL.sol";
+import "@aragon/os/contracts/apm/Repo.sol";
+import "@aragon/os/contracts/apm/APMNamehash.sol";
+import "@aragon/os/contracts/kernel/Kernel.sol";
+import "@aragon/os/contracts/lib/ens/ENS.sol";
+import "@aragon/os/contracts/lib/ens/PublicResolver.sol";
+import "@aragon/os/contracts/factory/DAOFactory.sol";
+import "@aragon/os/contracts/common/IsContract.sol";
+import "@aragon/id/contracts/IFIFSResolvingRegistrar.sol";
+
+
+contract BaseTemplate is APMNamehash, IsContract {
+    /* Hardcoded constant to save gas
+    * bytes32 constant internal AGENT_APP_ID = apmNamehash("agent");                  // agent.aragonpm.eth
+    * bytes32 constant internal VAULT_APP_ID = apmNamehash("vault");                  // vault.aragonpm.eth
+    * bytes32 constant internal VOTING_APP_ID = apmNamehash("voting");                // voting.aragonpm.eth
+    * bytes32 constant internal FINANCE_APP_ID = apmNamehash("finance");              // finance.aragonpm.eth
+    * bytes32 constant internal TOKEN_MANAGER_APP_ID = apmNamehash("token-manager");  // token-manager.aragonpm.eth
+    */
+
+    bytes32 constant internal AGENT_APP_ID = 0x9ac98dc5f995bf0211ed589ef022719d1487e5cb2bab505676f0d084c07cf89a;
+    bytes32 constant internal VAULT_APP_ID = 0x7e852e0fcfce6551c13800f1e7476f982525c2b5277ba14b24339c68416336d1;
+    bytes32 constant internal VOTING_APP_ID = 0x9fa3927f639745e587912d4b0fea7ef9013bf93fb907d29faeab57417ba6e1d4;
+    bytes32 constant internal FINANCE_APP_ID = 0xbf8491150dafc5dcaee5b861414dca922de09ccffa344964ae167212e8c673ae;
+    bytes32 constant internal TOKEN_MANAGER_APP_ID = 0x6b20a3010614eeebf2138ccec99f028a61c811b3b1a3343b6ff635985c75c91f;
+
+    ENS public ens;
+    DAOFactory public daoFactory;
+    MiniMeTokenFactory public miniMeFactory;
+    IFIFSResolvingRegistrar public aragonID;
+
+    event DeployDao(address dao);
+    event DeployToken(address token);
+    event InstalledApp(address appProxy, bytes32 appId);
+
+    constructor(DAOFactory _daoFactory, ENS _ens, MiniMeTokenFactory _miniMeFactory, IFIFSResolvingRegistrar _aragonID) public {
+        require(isContract(address(_ens)), "TEMPLATE_ENS_NOT_CONTRACT");
+        require(isContract(address(_daoFactory)), "TEMPLATE_DAO_FACTORY_NOT_CONTRACT");
```

Comment:
Perhaps for `aragonID` and `miniMeFactory` we can check if it's 0x0, otherwise require them to also be contracts.

Also, given that we'll be using this base in who-knows-how-big templates, we may want to optimize its bytecode size. We could put everything into one require with one error message here. It would also be nice to lift the error strings into constants :).

Reply:
I wasn't checking the `aragonID` and `miniMeFactory` cause they may not be used. Instead, we are doing a lazy check when trying to use them, i.e. before creating a token and before registering an ID.

Reply:
> Instead, we are doing a lazy check when trying to use them

I'd rather do the expensive check (the `isContract()` check) on initialization than on each token creation and ID registration. If we check in the constructor that it's either `0x0` or a contract, then afterwards we can just make sure its not `0x0` when doing either of those actions.